server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -96,46 +96,22 @@ public Builder dynamicTemplates(Collection<DynamicTemplate> templates) {
 
         @Override
         public RootObjectMapper build(ContentPath contentPath) {
-            return (RootObjectMapper) super.build(contentPath);
-        }
-
-        @Override
-        protected ObjectMapper createMapper(String name, String fullPath, Explicit<Boolean> enabled, Nested nested, Dynamic dynamic,
-                Map<String, Mapper> mappers, Version indexCreatedVersion) {
-            assert nested.isNested() == false;
-            return new RootObjectMapper(name, enabled, dynamic, mappers,
-                    runtimeFields == null ? Collections.emptyMap() : runtimeFields,
-                    dynamicDateTimeFormatters,
-                    dynamicTemplates,
-                    dateDetection, numericDetection, indexCreatedVersion);
-        }
-    }
-
-    /**
-     * Removes redundant root includes in {@link ObjectMapper.Nested} trees to avoid duplicate
-     * fields on the root mapper when {@code isIncludeInRoot} is {@code true} for a node that is
-     * itself included into a parent node, for which either {@code isIncludeInRoot} is
-     * {@code true} or which is transitively included in root by a chain of nodes with
-     * {@code isIncludeInParent} returning {@code true}.
-     */
-    public void fixRedundantIncludes() {
-       fixRedundantIncludes(this, true);
-    }
-
-    private static void fixRedundantIncludes(ObjectMapper objectMapper, boolean parentIncluded) {
-        for (Mapper mapper : objectMapper) {
-            if (mapper instanceof ObjectMapper) {
-                ObjectMapper child = (ObjectMapper) mapper;
-                Nested nested = child.nested();
-                boolean isNested = nested.isNested();
-                boolean includeInRootViaParent = parentIncluded && isNested && nested.isIncludeInParent();
-                boolean includedInRoot = isNested && nested.isIncludeInRoot();
-                if (includeInRootViaParent && includedInRoot) {
-                    nested.setIncludeInParent(true);
-                    nested.setIncludeInRoot(false);
+            contentPath.add(name);
+            Map<String, Mapper> mappers = new HashMap<>();
+            for (Mapper.Builder builder : mappersBuilders) {
+                Mapper mapper = builder.build(contentPath);
+                Mapper existing = mappers.get(mapper.simpleName());
+                if (existing != null) {
+                    mapper = existing.merge(mapper);
                 }
-                fixRedundantIncludes(child, includeInRootViaParent || includedInRoot);
+                mappers.put(mapper.simpleName(), mapper);
```

Comment:
we previously had quite some dance around reusing the merging logic, I think, between object mapper and root object mapper. I see the merging logic directly in RootObjectMapper now, is it the same logic that is also used in what is left of ObjectMapper?

Reply:
Yes, I ended up duplicating it in RootObjectMapper so that I could separate out the builder implementations.

Reply:
ok, shall we then have at least some common static method so that the code is not duplicated?

Reply:
I've added a shared protected method in ObjectMapper.Builder